flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/CheckpointBarrierUnaligner.java

```diff
@@ -92,306 +90,173 @@
+			new CheckpointException(CheckpointFailureReason.CHECKPOINT_DECLINED_INPUT_END_OF_STREAM));
 	}
 
-	private void resetPendingCheckpoint(long checkpointId) {
-		if (isCheckpointPending()) {
-			LOG.warn("{}: Received barrier or EndOfPartition(-1) {} before completing current checkpoint {}. " +
-					"Skipping current checkpoint.",
-				taskName,
-				checkpointId,
-				currentConsumedCheckpointId);
+	private void resetPendingCheckpoint() {
+		LOG.warn("{}: Received barrier or EndOfPartition(-1) before completing current checkpoint {}. " +
+				"Skipping current checkpoint.",
+			taskName,
+			currentCheckpointId);
 
-			hasInflightBuffers.entrySet().forEach(hasInflightBuffer -> hasInflightBuffer.setValue(false));
-			numBarrierConsumed = 0;
-		}
+		storeNewBuffers.entrySet().forEach(storeNewBuffer -> storeNewBuffer.setValue(false));
+		numBarriersReceived = 0;
 	}
 
 	@Override
 	public long getLatestCheckpointId() {
-		return currentConsumedCheckpointId;
+		return currentCheckpointId;
 	}
 
 	@Override
 	public String toString() {
-		return String.format("%s: last checkpoint: %d", taskName, currentConsumedCheckpointId);
+		return String.format("%s: last checkpoint: %d", taskName, currentCheckpointId);
 	}
 
 	@Override
 	public void close() throws IOException {
 		super.close();
-		threadSafeUnaligner.close();
-	}
-
-	@Override
-	public boolean hasInflightData(long checkpointId, InputChannelInfo channelInfo) {
-		if (checkpointId < currentConsumedCheckpointId) {
-			return false;
-		}
-		if (checkpointId > currentConsumedCheckpointId) {
-			return true;
-		}
-		return hasInflightBuffers.get(channelInfo);
-	}
-
-	@Override
-	public CompletableFuture<Void> getAllBarriersReceivedFuture(long checkpointId) {
-		return threadSafeUnaligner.getAllBarriersReceivedFuture(checkpointId);
-	}
-
-	@Override
-	public Optional<BufferReceivedListener> getBufferReceivedListener() {
-		return Optional.of(threadSafeUnaligner);
+		allBarriersReceivedFuture.cancel(false);
 	}
 
 	@Override
 	protected boolean isCheckpointPending() {
-		return numBarrierConsumed > 0;
-	}
-
-	@VisibleForTesting
-	int getNumOpenChannels() {
-		return threadSafeUnaligner.getNumOpenChannels();
-	}
-
-	@VisibleForTesting
-	ThreadSafeUnaligner getThreadSafeUnaligner() {
-		return threadSafeUnaligner;
+		return numBarriersReceived > 0;
 	}
 
-	private void notifyCheckpoint(CheckpointBarrier barrier) throws IOException {
-		// ignore the previous triggered checkpoint by netty thread if it was already canceled or aborted before.
-		if (barrier.getId() >= threadSafeUnaligner.getCurrentCheckpointId()) {
-			super.notifyCheckpoint(barrier, 0);
+	@Override
+	public void processBuffer(Buffer buffer, InputChannelInfo channelInfo) {
+		if (storeNewBuffers.get(channelInfo)) {
+			checkpointCoordinator.getChannelStateWriter().addInputData(
+				currentCheckpointId,
+				channelInfo,
+				ChannelStateWriter.SEQUENCE_NUMBER_UNKNOWN,
+				ofElement(buffer.retainBuffer(), Buffer::recycleBuffer));
```

Comment:
Here, we are persisting in-flight buffers, only as they are being processed? Doesn't it mean, that unaligned checkpoint will be completed only after we process all of the buffers, making unaligned checkpoint just as quick as aligned?

Reply:
Let me sketch the most general data flow:
* We have couple of buffers in our input channel.
* Checkpoint is triggered by another channel. *
* All buffers that are now pulled by `CheckpointedInputGate` from the first channel get persisted by above function.
* Then the barrier comes in. *
* It overtakes all buffers and is now at the head. *
* `CheckpointedInputGate` gets priority notification and polls the barrier.
* Upon dispatching, it calls `Unaligner`, which spills additionally all overtaken buffers.
* Further buffers are not persisted.

All steps marked with * are performed in a different thread (other task thread / netty).

For me this is equivalent to the current behavior, but maybe I missed something.